### PR TITLE
EVG-2876 create route to save subscription preferences

### PR DESCRIPTION
**Pull request by john-m-liu**

### CODE REVIEW

File: model/user/user.go [hunk 1]
```diff
@@ -35,9 +36,16 @@ type PubKey struct {
 }
 
 type UserSettings struct {
-	Timezone     string     `json:"timezone" bson:"timezone"`
-	NewWaterfall bool       `json:"new_waterfall" bson:"new_waterfall"`
-	GithubUser   GithubUser `json:"github_user" bson:"github_user,omitempty"`
+	Timezone      string                  `json:"timezone" bson:"timezone"`
+	NewWaterfall  bool                    `json:"new_waterfall" bson:"new_waterfall"`
+	GithubUser    GithubUser              `json:"github_user" bson:"github_user,omitempty"`
+	SlackUsername string                  `bson:"slack_username" json:"slack_username"`
```
Comment (richardsamuels) [inline model/user/user.go]: omitempty on this and notifications?

File: model/user/user.go [hunk 2]
```diff
@@ -35,9 +36,16 @@ type PubKey struct {
...
+	Notifications NotificationPreferences `bson:"notifications" json:"notifications"`
+}
+
+type NotificationPreferences struct {
+	BuildBreak  event.UserSubscriptionPreference `bson:"build_break" json:"build_break"`
```
Comment (richardsamuels) [inline model/user/user.go]: omitempty on these as well?

File: model/event/subscribers.go [hunk 3]
```diff
@@ -17,6 +17,15 @@ const (
...
 	SlackSubscriberType             = "slack"
+	NoSubscriberType                = "none"
+)
+
+type UserSubscriptionPreference string
+
+const (
+	PreferenceEmail UserSubscriptionPreference = EmailSubscriberType
+	PreferenceSlack                            = SlackSubscriberType
+	PreferenceNone                             = NoSubscriberType
```
Comment (richardsamuels) [inline model/event/subscribers.go]: This vs empty string. Thoughts? Also, I feel this belongs in model.user

File: model/event/subscribers.go [hunk 4]
```diff
@@ -17,6 +17,15 @@ const (
 	EvergreenWebhookSubscriberType  = "evergreen-webhook"
 	EmailSubscriberType             = "email"
 	SlackSubscriberType             = "slack"
+	NoSubscriberType                = "none"
```
Comment (richardsamuels) [inline model/event/subscribers.go]: I don't like the idea of NoSubscriberType because it's a non-zero type that is technically invalid (i.e. it's not a subscriber"). If my memory is correct, the subscriber model will also fail to unmarshal something with nil data.

File: model/event/subscribers.go [hunk 5]
```diff
@@ -133,3 +142,12 @@ type GithubPullRequestSubscriber struct {
 func (s *GithubPullRequestSubscriber) String() string {
 	return fmt.Sprintf("%s-%s-%d-%s", s.Owner, s.Repo, s.PRNumber, s.Ref)
 }
+
+func IsValidSubscriptionPreference(in string) bool {
```
Comment (richardsamuels) [inline model/event/subscribers.go]: See note above about model.user

File: model/event/subscribers.go (hunk 4 above)
Comment (john-m-liu) [reply model/event/subscribers.go]: Fixed. I initially thought I would need a way to differentiate between "no change" and "no notifications" in the route but that's not needed

File: model/event/subscribers.go (hunk 3 above)
Comment (john-m-liu) [reply model/event/subscribers.go]: Removed + moved

File: model/event/subscribers.go (hunk 5 above)
Comment (john-m-liu) [reply model/event/subscribers.go]: moved

File: model/user/user.go (hunk 1 above)
Comment (john-m-liu) [reply model/user/user.go]: done

File: model/user/user.go (hunk 2 above)
Comment (john-m-liu) [reply model/user/user.go]: I feel like these ones shouldn't be omitempty if the notification prefs are omitempty

File: model/user/user.go [hunk 6]
```diff
@@ -159,3 +174,12 @@ func (u *DBUser) IncPatchNumber() (int, error) {
 	return dbUser.PatchNumber, nil
 
 }
+
+func IsValidSubscriptionPreference(in string) bool {
+	switch in {
+	case event.EmailSubscriberType, event.SlackSubscriberType:
```
Comment (bsamek) [inline model/user/user.go]: nit, `case in PreferenceEmail, PreferenceSlack:`, since you've redefined them locally?

File: rest/model/user.go [hunk 7]
```diff
@@ -26,3 +27,108 @@ func (apiPubKey *APIPubKey) BuildFromService(h interface{}) error {
...
+	SlackUsername APIString                  `json:"slack_username"`
+	Notifications APINotificationPreferences `json:"notifications"`
+}
+
+func (s *APIUserSettings) BuildFromService(h interface{}) error {
+	switch v := h.(type) {
+	case user.UserSettings:
+		s.Timezone = ToAPIString(v.Timezone)
+		s.SlackUsername = ToAPIString(v.SlackUsername)
+		err := s.GithubUser.BuildFromService(v.GithubUser)
```
Comment (bsamek) [inline rest/model/user.go]: I was imagining that GithubUser could be empty, if the user hadn't set it yet. Does this noerror in that case?

File: rest/model/user_test.go [hunk 8]
```diff
@@ -0,0 +1,32 @@
+package model
+
+import (
+	"testing"
+
+	"github.com/evergreen-ci/evergreen/model/user"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestUserSettings(t *testing.T) {
```
Comment (bsamek) [inline rest/model/user_test.go]: Add a test that validates the empty settings case/partial settings case.

File: model/user/user.go (hunk 6 above)
Comment (john-m-liu) [reply model/user/user.go]: They're different types. If I have to do that I have to type assert to string, which is messier

File: rest/model/user.go (hunk 7 above)
Comment (john-m-liu) [reply rest/model/user.go]: If that property is empty, this will also return an empty GithubUser

File: rest/model/user_test.go (hunk 8 above)
Comment (john-m-liu) [reply rest/model/user_test.go]: added

**Comment by richardsamuels** [comment]
@john-m-liu Can you also post a screen shot please?

**Comment by john-m-liu** [comment]
This is a rest route. What did you want a screenshot of?
